astrolabe/commands.py

```diff
@@ -0,0 +1,138 @@
+from time import time, sleep
+
+from astrolabe.utils import assert_subset
+from atlasclient import AtlasApiError
+
+
+def get_one_organization_by_name(*, client, organization_name):
+    all_orgs = client.orgs.get().data
+    for org in all_orgs.results:
+        if org.name == organization_name:
+            return org
+    raise AtlasApiError('Resource not found.')
+
+
+def ensure_project(*, client, group_name, organization_id):
+    try:
+        return client.groups.post(
+            name=group_name, orgId=organization_id).data
+    except AtlasApiError as exc:
+        if exc.error_code == 'GROUP_ALREADY_EXISTS':
+            return client.groups.byName[group_name].get().data
+        else:
+            raise
+
+
+def ensure_admin_user(*, client, group_id, username, password):
+    user_details = {
+        "groupId": group_id,
+        "databaseName": "admin",
+        "roles": [{
+            "databaseName": "admin",
+            "roleName": "atlasAdmin"}],
+        "username": username,
+        "password": password}
+
+    try:
+        return client.groups[group_id].databaseUsers.post(**user_details).data
+    except AtlasApiError as exc:
+        if exc.error_code == "USER_ALREADY_EXISTS":
+            username = user_details.pop("username")
+            return client.groups[group_id].databaseUsers.admin[username].patch(
+                **user_details).data
+        else:
+            raise
+
+
+def ensure_connect_from_anywhere(*, client, group_id,):
+    ip_details_list = [{"cidrBlock": "0.0.0.0/0"}]
+    client.groups[group_id].whitelist.post(json=ip_details_list)
+
+
+def get_cluster_state(client, group_name, cluster_name):
+    project = client.groups.byName[group_name].get().data
+    cluster = client.groups[project.id].clusters[cluster_name].get().data
+    return cluster.stateName
+
+
+def is_cluster_state(client, group_name, cluster_name, target_state):
+    project = client.groups.byName[group_name].get().data
+    cluster = client.groups[project.id].clusters[cluster_name].get().data
+    return cluster.stateName == target_state
+
+
+def wait_until_cluster_state(client, group_name, cluster_name, target_state,
+                             polling_frequency, polling_timeout):
+    if is_cluster_state(client, group_name, cluster_name,
+                        target_state):
+        return True
+
+    start_time = time()
+    sleep_interval = 1 / polling_frequency
+    while (time() - start_time) < polling_timeout:
+        if is_cluster_state(client, group_name, cluster_name, target_state):
+            return True
+        sleep(sleep_interval)
+
+    return False
+
+
+def select_callback(callback, args, kwargs, frequency, timeout):
```

Comment:
It seems like select_callback, wait_until_cluster_state, and BooleanCallableSelector all call a method every X seconds until that method returns true. Can they all use the same loop?

Reply:
This was old code. Please disregard and take another look. The "selector" is now only implemented in `poller.py` (might be in need of a name change).